### PR TITLE
Implement atoms=false for estimated_buy_amount

### PR DESCRIPTION
**Pull request by e00E**
and add error code handling of rejections

First step for #1082 . So far this commit only implements this for one route but not the others. If this way of implementation makes sense I will add it to the other routes (should it apply to all routes, even `/marktets`?) and update the readme. Then we probably also want the token info overrides like the driver has for tokens that don't implement ERC20 properly.

### Test Plan
Unit test for error codes and manual test:
```
$ curl "localhost:8080/api/v1/markets/4-7/estimated-buy-amount/1000000000000000000?atoms=true"
{"baseTokenId":"4","quoteTokenId":"7","buyAmountInBase":"1006194.8387101036","sellAmountInQuote":"1000000000000000000"}

$ curl "localhost:8080/api/v1/markets/4-7/estimated-buy-amount/1?atoms=false"
{"baseTokenId":"4","quoteTokenId":"7","buyAmountInBase":"1.0061948387101036","sellAmountInQuote":"1"}
```
Tested that changing the type of token amounts from u128 to f64 doesn't cause problems in frontend.

### CODE REVIEW

File: price-estimator/src/filter.rs [hunk 1]
```diff
@@ -127,20 +168,43 @@ async fn get_markets<T>(
...
     orderbook: Arc<Orderbook<T>>,
-) -> Result<impl warp::Reply, Infallible> {
+    token_infos: Arc<dyn TokenInfoFetching>,
+) -> Result<impl Reply, Rejection> {
+    let sell_amount_in_quote_atoms = if query.atoms {
+        sell_amount_in_quote
+    } else {
+        let token_info = get_token_info(token_pair.sell_token_id, token_infos.as_ref()).await?;
+        sell_amount_in_quote * token_info.unit_in_atoms()
+    };
```
Comment (fleupold) [inline price-estimator/src/filter.rs:194]: I was wondering if this logic should rather live in the estimate_buy_amount_filter (so that the actual handler can only deal with atom based logic) but that doesn't nicely work with the way that results are returned at the moment since they also have to be changed if `atoms = false`. 
Given that the handler itself just delegates a call into the pricegraph, maybe it's fine to keep this logic here.

File: price-estimator/src/models.rs [hunk 2]
```diff
@@ -62,9 +62,9 @@ pub struct EstimatedOrderResult {
     #[serde(with = "display_fromstr")]
     pub quote_token_id: u16,
     #[serde(with = "display_fromstr")]
-    pub buy_amount_in_base: u128,
+    pub buy_amount_in_base: f64,
     #[serde(with = "display_fromstr")]
-    pub sell_amount_in_quote: u128,
+    pub sell_amount_in_quote: f64,
```
Comment (fleupold) [inline price-estimator/src/models.rs:67]: Does this work with the frontend in the case that atoms=true?

File: price-estimator/src/filter.rs [hunk 3]
```diff
@@ -209,9 +273,78 @@ fn apply_rounding_buffer(amount: f64, price_rounding_buffer: f64) -> f64 {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use futures::future::FutureExt as _;
+    use anyhow::{anyhow, Result};
+    use core::orderbook::StableXOrderBookReading;
+    use futures::future::{BoxFuture, FutureExt as _};
+
+    fn empty_orderbook() -> impl StableXOrderBookReading {
```
Comment (fleupold) [inline price-estimator/src/filter.rs:280]: Is there a reason you are not using the Mocked versions here?

File: core/src/token_info.rs [hunk 4]
```diff
@@ -51,6 +51,11 @@ impl TokenBaseInfo {
         SYMBOL_OVERRIDES.get(&self.alias).unwrap_or(&self.alias)
     }
 
+    /// One unit of the token taking decimals into account, given in number of atoms.
+    pub fn unit_in_atoms(&self) -> f64 {
```
Comment (fleupold) [inline core/src/token_info.rs:55]: ubernit: maybe base_unit instead of just unit (since technically an atom is also a unit)

File: price-estimator/src/models.rs (hunk 2 above)
Comment (e00E) [reply price-estimator/src/models.rs:67]: Either way all json numbers can be fractions and something that previously wasn't a fraction as u128 still won't be a fraction as f64 so I didn't expect this to be a problem but I'll test it.

File: price-estimator/src/filter.rs (hunk 3 above)
Comment (e00E) [reply price-estimator/src/filter.rs:280]: Can't use mocks in other crates because they are only compiled for tests.

File: price-estimator/src/models.rs [hunk 5]
```diff
@@ -130,6 +130,11 @@ impl From<&pricegraph::TransitiveOrderbook> for MarketsResult {
 #[serde(transparent)]
 pub struct PriceEstimateResult(pub Option<f64>);
 
+#[derive(Debug, Serialize)]
+pub struct ErrorMessage {
+    pub message: &'static str,
+}
```
Comment (nlordell) [inline price-estimator/src/models.rs:136]: nit: I think the struct shouldn't be an `ErrorMessage` but more of an `Error`. Also I think the `message` field should be used for a human readable messages that may include some extra context and not status constants (which should probably be an enum instead of `&'static str`).
```suggestion
#[derive(Debug, Serialize)]
pub struct Error {
    pub status: ErrorStatus,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorState {
    NotFound,
    NoTokenInfo,
    // ...
}
```

File: price-estimator/src/filter.rs (hunk 1 above)
Comment (nlordell) [reply price-estimator/src/filter.rs:194]: Maybe it can be implemented in the `filter` by using a `map` function?

File: price-estimator/src/main.rs [hunk 6]
```diff
@@ -127,3 +133,18 @@ fn setup_driver_metrics() -> HttpMetrics {
     });
     HttpMetrics::new(&prometheus_registry).unwrap()
 }
+
+async fn initialize_token_info_cache(cache: &TokenInfoCache) -> anyhow::Result<()> {
+    for token_id in cache.all_ids().await.context("failed to get all ids")? {
+        // Individual tokens might not conform to erc20 in which case we are unable to retrieve
+        // their info.
+        if let Err(err) = cache.get_token_info(token_id).await {
```
[the remark this answers is not in the document]
Comment (nlordell) [reply price-estimator/src/main.rs:141]: :+1: I think it should be a `all_token_infos` method on the `TokenInfoFetching` trait that provides a default implementation.

File: price-estimator/src/models.rs (hunk 5 above)
Comment (e00E) [reply price-estimator/src/models.rs:136]: I disagree. This struct represents the body of the response we send when we do not respond with success. The `handle_rejection` function decides what goes in there so I see no reason to introduce an enum with more complicated serialization.
I agree with the human readable part so i'll change the messages a bit to contain more informatin.

File: price-estimator/src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (e00E) [reply price-estimator/src/main.rs:141]: Will do this in another PR.

File: price-estimator/src/models.rs (hunk 2 above)
Comment (e00E) [reply price-estimator/src/models.rs:67]: Tested it without problems.

File: price-estimator/src/filter.rs (hunk 1 above)
Comment (e00E) [reply price-estimator/src/filter.rs]: I don't think this will turn out better by doing at a higher level so I'll keep it this way. I'm not opposed to that in the future either just don't think it's worth trying for this PR.

**Comment by e00E** [comment]
Ready for review now. I've included a readme update as well.

File: price-estimator/src/models.rs (hunk 5 above)
Comment (nlordell) [reply price-estimator/src/models.rs:130]: My main argument was that if the error message was going to be `NOT_FOUND` style string constants, then it should be an enum (and a `status` or `kind` as opposed to a `message`). This could be very valuable for clients to recover an exact error type from the error data (kind of like the `ErrorKind` pattern in Rust).

`messages` are intended to be human readable error messages, kind of like `Error::description` or `Error::to_string` in Rust.

File: price-estimator/src/models.rs (hunk 2 above)
Comment (nlordell) [reply price-estimator/src/models.rs:61]: I think, **in a future PR**, we should change these to serialize as numbers instead of strings.

Comment (e00E) [reply price-estimator/src/models.rs:61]: Agree, I believe we have an issue for this.

File: price-estimator/src/models.rs (hunk 5 above)
Comment (e00E) [reply price-estimator/src/models.rs:130]: I'm not sure if we should make any guarantees about the contents of the error. Human readable has practical benefits but I don't think any client can recover in a meaningful way so I don't want to make more work for us to implement this. This is something I'm not super confident about so I could be convinced but I would also like to have that discussion in another issue/pr and keep this simple way of handling errors for now.

Comment (nlordell) [reply price-estimator/src/models.rs:130]: > I would also like to have that discussion in another issue/pr

:+1: 

I just have a nit with the naming, maybe
```suggestion
#[derive(Debug, Serialize)]
pub struct ErrorResult {
    pub message: &'static str,
}
```

File: price-estimator/README.md [hunk 7]
```diff
@@ -1,18 +1,25 @@
 ## API
 
+All endpoints use the query part of the url with these key-values:
+
+* `atoms`: Required. If set to `true` all amounts are denominated in the smallest available unit (base quantity) of the token. If `false` all amounts are denominated in the "natural" unit of the respective token given by the number of decimals specified through the ERC20 interface. TODO: `false` is currently only implemented for estimated-buy-amount .
```
Comment (fleupold) [inline price-estimator/README.md:5]: I think it would make sense to default this to false (can be a separate PR given the size of this one).

File: price-estimator/src/models.rs (hunk 5 above)
Comment (fleupold) [reply price-estimator/src/models.rs:130]: One reason error kinds are nice is that they frontend can easily localize them to other languages (which a backend API usually doesn't). But I agree that it is not needed in our current state.
